working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,612 @@
+
+To rename the class reference parameter, we select the `validate_model` function again, open the inline Copilot chat, and provide the following prompt:
+
+> Please use cls for the class parameter name
+
+Find & Replace could have done this just as easily, and we might have had more confidence in the results, but changing the rest of the code to use the refactored function is a bit more complex. Let's try having Copilot helping us with the rest of the change. 
+
+One approach we can try is to open each file, select the entire contents, and ask Copilot to change any instances of `validate_book` to `validate_model`. We need to carefully review the lines Copilot suggests changing to ensure nothing unexpected is altered, but it can save us a little time since we need to do slightly more than find and replace the function name.
+
+Due to differences in context, it's possible that a prompt that works fine in one file may not work as well in another file. For example, in the `routes.py` file, the following prompt, which provides very few details, is able to produce the desired changes:
+
+> Please replace all uses of validate_book with validate_model
+
+However, in the `test_routes.py` file, more detail is necessary for Copilot to recognize that there is a new class parameter which is required wherever the function is invoked, and that the `Book` class needs to be imported:
+
+> Please replace all uses of validate_book(book_id) with validate_model(cls, model_id). Import any required model classes.
+
+Since there is always a degree of validation and review required with any change when using Copilot, if we're having trouble finding an effective prompt, it may be worth doing the refactoring some other way. We can always refactor by hand, or by using dedicated refactoring utilities or those built into a development environment. Copilot is just one tool among many that we can draw on as developers.
+
+## Summary
+
+This hands-on experience in a project wraps up our whirlwind tour of GitHub's Copilot! Whether we're writing new code or working in an existing codebase, Copilot can help us with tasks like generating ideas for tests, surfacing error handling we may have missed, and generally writing code more quickly. Even when Copilot's suggestions look good at a glance, we always need to carefully proofread them both for correctness and completeness. Just because a code suggestion works does not mean it fits all of our needs, so we need to remain vigilant and critical of what Copilot provides.
+
+If you're interested in more practice with Copilot, we highly suggest revisiting Solar System, Task List, Inspiration Board, or any other project that you're already familiar with so you know what the end product could look like and can focus on working with Copilot.
```

Comment:
- rephrased to go you-less
```suggestion
For more practice with Copilot, think about revisiting Solar System, Task List, Inspiration Board, or any other familiar project. When we already have a reasonable idea what the end product could look like, that let's us focus on practicing new skills, such as working with Copilot.
```

Reply:
Updating suggestion to remove contraction before adding to batch.

```suggestion
For more practice with Copilot, think about revisiting Solar System, Task List, Inspiration Board, or any other familiar project. When we already have a reasonable idea what the end product could look like, that lets us focus on practicing new skills, such as working with Copilot.
```